### ISSUE TITLE
use socket.io in project,  running build command will report an errorwrong: Docusaurus Node/SSR could not render static page with path=/ because of error: Buffer is not defined

### ISSUE THREAD

**Issue by Feafortecsy** [facebook/docusaurus#4836]
## online demo  to reproduce

[codesandbox](https://codesandbox.io/s/suspicious-rosalind-pfzqc),  open this page and run build in terminal

i use [socket.io](https://github.com/socketio/socket.io) in my project,  and local development is normal, but running the yarn build command will report an errorwrong

![20210522130634](https://user-images.githubusercontent.com/31724313/119215180-93db1480-bafe-11eb-91b9-f9e9362796bd.png)

## steps

1. Add socket. IO to your project's package.json and install it,
2. use sockets in your project.
3. yarn build

## Expected behavior

build my project successfully

## environment

windows 10 20h2
``` json
{
  "bufferutil": "4.0.3",
  "socket.io": "4.1.2",
  "socket.io-client": "4.1.2",
  "@docusaurus/core": "2.0.0-beta.0",
  "@docusaurus/preset-classic": "2.0.0-beta.0",
  "utf-8-validate": "5.0.5"
}
```

**Comment by Josh-Cena** [comment]
lol, just resolved quite the same issue on Discord. This is because Buffer is a browser API and can't be accessed in server-side rendering. This is typically resolved by wrapping the entire code that used this API (e.g. `socket.io-client`) in `React.useEffect`, but I'd like to see if there're other ways to trigger a client-side rendering...

**Comment by Feafortecsy** [comment]
> lol, just resolved quite the same issue on Discord. This is because Buffer is a browser API and can't be accessed in server-side rendering. I resolved mine by wrapping the entire code that used this API (e.g. `socket.io-client`) in `React.useEffect`, but I'd like to see if there're other ways to trigger a client-side rendering...

Thank you for your reply.

It should be noted that `Socket. IO` has already divided the server side and the client side `Socket.io-client`. The server side should have been compatible, so I think it is probably the `WS` library Socket. IO uses. However, I also hope that `Docusaurus` exposes some APIs that will allow us to customize some Webpack configurations, thus avoiding a few bugs

**Comment by Feafortecsy** [comment]
> wrapping the entire code that used this API (e.g. `socket.io-client`) in `React.useEffect`

About what you said。 This error appears in the server code. I do not know how to use the hook of useEffect. Could you please give me an example code

**Comment by Josh-Cena** [comment]
For example, I was writing a component that uses `localStorage` to store its state. Previously I wrote:

```js
function component() {
  const [state, setState] = useState(0);
  const history = localStorage.getItem("data");
  if (history !== null) setState(JSON.parse(history));
  return <div>{state}</div>;
}
```

which results in `localStorage is not defined` during SSR. In the end I wrapped it in `useEffect`:

```js
function component() {
  const [state, setState] = useState(0);
  useEffect(() => {
    const history = localStorage.getItem("data");
    if (history !== null) setState(JSON.parse(history));
  }, []);
  return <div>{state}</div>;
}
```

Which delays the execution until runtime (the empty deps list makes it equivalent to `componentDidMount` and will only be run once) and it was built successfully.

A similar usecase here with a package API: https://github.com/vercel/next.js/issues/11619

But I'm not sure if it's the most elegant solution, especially with `import` statements—but it works anyways. I'd wait for someone else to give a better solution 👀

**Comment by Feafortecsy** [comment]
> For example, I was writing a component that uses `localStorage` to store its state. Previously I wrote:
> 
> ```js
> function component() {
>   const [state, setState] = useState(0);
>   const history = localStorage.getItem("data");
>   if (history !== null) setState(JSON.parse(history));
>   return <div>{state}</div>;
> }
> ```
> 
> which results in `localStorage is not defined` during SSR. In the end I wrapped it in `useEffect`:
> 
> ```js
> function component() {
>   const [state, setState] = useState(0);
>   useEffect(() => {
>     const history = localStorage.getItem("data");
>     if (history !== null) setState(JSON.parse(history));
>   }, []);
>   return <div>{state}</div>;
> }
> ```
> 
> Which delays the execution until runtime (the empty deps list makes it equivalent to `componentDidMount` and will only be run once) and it was built successfully.
> 
> A similar usecase here with a package API: [vercel/next.js#11619](https://github.com/vercel/next.js/issues/11619)
> 
> But I'm not sure if it's the most elegant solution, especially with `import` statements—but it works anyways. I'd wait for someone else to give a better solution 👀

For other reasons, I gave up this library in order to ensure that the project is online. Thanks for the help!!!

**Comment by lex111** [comment]
I am going to close this issue as it is not Docusaurus bug, however we will improve SSR error message to help the user fix it themselves (see screenshot in #4875).
